### PR TITLE
Optimize Type-dependency operations

### PR DESCRIPTION
**Pull request by SergeiPavlov**
This is subset of https://github.com/DataObjects-NET/dataobjects-net/pull/242 

Improvements:
* Get rid of intermediate dictionaries `TypeInfo` -> `Interfaces/Descendants/Ancestors`.  These lists/sets are now fields of `TypeInfo`
* Split `GetInterfaces/GetAncestors(recursive)` into properties: `Interfaces`/`RecursiveInterfaces` ... to avoid unnecessary branching baceuse caller side invokes the methods with constant argument

### CODE REVIEW

File: Orm/Xtensive.Orm/Orm/Building/Builders/StorageMappingValidator.cs [hunk 1]
```diff
@@ -69,7 +69,7 @@ private void EnsureMappingDatabaseIsValid()
     private void EnsureIntefacesAreImplementedWithinSingleDatabase()
     {
       foreach (var @interface in model.Types.Where(t => t.IsInterface)) {
-        var implementors = @interface.GetImplementors(true).ToList();
+        var implementors = @interface.RecursiveImplementors.ToList();
```
Comment (botinko) [inline Orm/Xtensive.Orm/Orm/Building/Builders/StorageMappingValidator.cs:72]: is `ToList()` is redundant?

File: Orm/Xtensive.Orm/Collections/Interfaces/IFilterable.cs [hunk 2]
```diff
@@ -21,14 +21,14 @@ public interface IFilterable<TFilter, TItem>
     /// </summary>
     /// <param name="criteria">The object to filter initial collection with.</param>
     /// <returns><see cref="ICollection{TItem}"/> object.</returns>
```
Comment (botinko) [inline Orm/Xtensive.Orm/Collections/Interfaces/IFilterable.cs:23]: need to update comments

File: Orm/Xtensive.Orm/Collections/Interfaces/IFilterable.cs [hunk 3]
```diff
@@ -21,14 +21,14 @@ public interface IFilterable<TFilter, TItem>
...
+    IEnumerable<TItem> Find(TFilter criteria);
 
     /// <summary>
     /// Finds the items from initial collection according to specified filter <paramref name="criteria"/>.
     /// </summary>
     /// <param name="criteria">The object to filter initial collection with.</param>
     /// <param name="matchType">Type of the match.</param>
     /// <returns><see cref="ICollection{TItem}"/> object.</returns>
-    ICollection<TItem> Find(TFilter criteria, MatchType matchType);
+    IEnumerable<TItem> Find(TFilter criteria, MatchType matchType);
```
Comment (botinko) [inline Orm/Xtensive.Orm/Collections/Interfaces/IFilterable.cs:32]: just wondering what is the reason for this change?

Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Collections/Interfaces/IFilterable.cs:32]: The safety of interface
`ICollection` is modifiable, but `.Find()` callers need only to enumerate results

**Review by botinko — APPROVED** [review]

File: Orm/Xtensive.Orm/Collections/Interfaces/IFilterable.cs (hunk 2 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Collections/Interfaces/IFilterable.cs]: fixed

File: Orm/Xtensive.Orm/Orm/Building/Builders/StorageMappingValidator.cs (hunk 1 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Building/Builders/StorageMappingValidator.cs]: yes, fixed
